src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs

```diff
@@ -10,299 +11,63 @@ namespace System.Text.Json.Serialization.Converters
     /// <summary>
     /// Default base class implementation of <cref>JsonDictionaryConverter{TCollection}</cref> .
     /// </summary>
-    internal abstract class DictionaryDefaultConverter<TCollection, TKey, TValue>
-        : JsonDictionaryConverter<TCollection>
+    internal abstract class DictionaryDefaultConverter<TDictionary, TKey, TValue>
+        : JsonDictionaryConverter<TDictionary, TKey, TValue>
+        where TDictionary : IEnumerable<KeyValuePair<TKey, TValue>>
         where TKey : notnull
     {
-        /// <summary>
-        /// When overridden, adds the value to the collection.
-        /// </summary>
-        protected abstract void Add(TKey key, in TValue value, JsonSerializerOptions options, ref ReadStack state);
-
-        /// <summary>
-        /// When overridden, converts the temporary collection held in state.Current.ReturnValue to the final collection.
-        /// This is used with immutable collections.
-        /// </summary>
-        protected virtual void ConvertCollection(ref ReadStack state, JsonSerializerOptions options) { }
-
-        /// <summary>
-        /// When overridden, create the collection. It may be a temporary collection or the final collection.
-        /// </summary>
-        protected virtual void CreateCollection(ref Utf8JsonReader reader, ref ReadStack state) { }
-
-        internal override Type ElementType => typeof(TValue);
-
-        internal override Type KeyType => typeof(TKey);
-
-
-        protected JsonConverter<TKey>? _keyConverter;
-        protected JsonConverter<TValue>? _valueConverter;
-
-        protected static JsonConverter<T> GetConverter<T>(JsonTypeInfo typeInfo)
-        {
-            JsonConverter<T> converter = (JsonConverter<T>)typeInfo.PropertyInfoForTypeInfo.ConverterBase;
-            Debug.Assert(converter != null); // It should not be possible to have a null converter at this point.
-
-            return converter;
-        }
-
-        internal sealed override bool OnTryRead(
-            ref Utf8JsonReader reader,
-            Type typeToConvert,
+        protected internal override bool OnWriteResume(
+            Utf8JsonWriter writer,
+            TDictionary value,
             JsonSerializerOptions options,
-            ref ReadStack state,
-            [MaybeNullWhen(false)] out TCollection value)
+            ref WriteStack state)
         {
-            JsonTypeInfo elementTypeInfo = state.Current.JsonTypeInfo.ElementTypeInfo!;
-
-            if (state.UseFastPath)
+            IEnumerator<KeyValuePair<TKey, TValue>> enumerator;
```

Comment:
One reason for the one-off code is to use built-in value enumerators (e.g. `Dictionary<TKey,TValue>.Enumerator`) to avoid boxing of primitives such as values.

Reply:
The converters that special case enumeration still override the method. The converters reusing the base implementation are specifically the ones that replicate the same implementation line-by-line.

Still, I ran the benchmarks but as expected there were no regressions recorded.

Reply:
I see -- the interface-based ones like `IDictionary<,>` use the shared code now, but not the converter for `Dictionary<,>`.

Basically we need to be sure value-type-based enumerators are not boxed (including `Dictionary<TKey,TValue>.Enumerator`) especially for each element being enumerated (we don't care about boxing for the continuation case) and of course the keys\values are not boxed.

Also the converter for `System.Array` is another special converter that uses indexes instead of allocating an enumerator, but there were no real changes here for that.